File: coco_scala/src/main/scala/exp/m3/M3.scala
package exp.m3

trait Factory[T] extends exp.m2.Factory[T] {
  def mult(left : exp.Exp[T], right : exp.Exp[T]) : exp.Exp[T]
  def neg(inner : exp.Exp[T]) : exp.Exp[T]
  def divd(left : exp.Exp[T], right : exp.Exp[T]) : exp.Exp[T]
}

trait Mult[T] extends exp.m2.Exp[T] {
  def left : exp.Exp[T]
  def right : exp.Exp[T]

  def eval : Double = left.eval * right.eval
  def prettyp : String = String.format("(%s * %s)", left.prettyp, right.prettyp)
}

trait Neg[T] extends exp.m2.Exp[T] {
  def inner : exp.Exp[T]

  def eval : Double = -inner.eval
  def prettyp : String = String.format("(- %s)", inner.prettyp)
}

trait Divd[T] extends exp.m2.Exp[T] {
  def left: exp.Exp[T]
  def right: exp.Exp[T]

  def eval: Double = left.eval / right.eval
  def prettyp: String = String.format("(%s / %s)", left.prettyp, right.prettyp)
}

object finalized {
  import exp.m2.finalized.Exp

  trait Factory extends exp.m3.Factory[Exp] with exp.m2.finalized.Factory {
    override def mult(left: exp.Exp[Exp], right: exp.Exp[Exp]):Exp = new Mult(left, right)
    override def neg(inner: exp.Exp[Exp]):Exp = new Neg(inner)
    override def divd(left: exp.Exp[Exp], right: exp.Exp[Exp]):Exp = new Divd(left, right)
  }

  class Mult(val left:exp.Exp[Exp], val right:exp.Exp[Exp]) extends exp.m3.Mult[Exp] with Exp
  class Neg(val inner:exp.Exp[Exp]) extends exp.m3.Neg[Exp] with Exp
  class Divd(val left:exp.Exp[Exp], val right:exp.Exp[Exp]) extends exp.m3.Divd[Exp] with Exp
}

File: coco_csharp/CoCo.Tests/M3.cs
using System;
using Xunit;
using EP.M3;

namespace CoCo.Tests.M3
{
    public interface TestTemplate<FT> : CoCo.Tests.M2.TestTemplate<FT>, Factory<FT> {
        new public void test()
        {
            CoCo.Tests.M2.TestTemplate<FT> parent = this;
            parent.test();

            EP.Exp<FT> expr1 = this.mult(this.lit(3.0), this.lit(2.0));
            Assert.Equal("(3*2)", convert(expr1).prettyp());

            EP.Exp<FT> expr2 = this.divd(this.lit(3.0), this.lit(2.0));
            Assert.Equal("(3/2)", convert(expr2).prettyp());

            EP.Exp<FT> expr3 = this.neg(this.lit(3.0));
            Assert.Equal("(-3)", convert(expr3).prettyp());

            EP.Exp<FT> expr4 = this.neg(expr1);
            Assert.Equal("(-(3*2))", convert(expr4).prettyp());

            Assert.Equal(convert(this.neg(this.lit(1.0))).eval(), -1.0);
            Assert.Equal("(-1)", convert(this.neg(this.lit(1.0))).prettyp());
            Assert.Equal("((5/2)*4)", convert(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0))).prettyp());
            Assert.Equal(10, convert(this.mult(this.divd(this.lit(5.0), this.lit(2.0)), this.lit(4.0))).eval());
            Assert.Equal(convert(this.neg(this.lit(5.0))).eval(), -5.0);
            Assert.Equal("(-(2*3))", convert(this.neg(this.mult(this.lit(2.0), this.lit(3.0)))).prettyp());
        }
    }

    public class ActualTest : TestTemplate<EP.M2.Finalized.Exp>, EP.M3.Finalized.Factory {}

    public class M3Test
    {
        [Fact]
        public void TestM3()
        {
            TestTemplate<EP.M2.Finalized.Exp> theTest = new ActualTest();
            theTest.test();
        }
    }
}
